File: src/main/kotlin/com/mobaassistant/service/MetaAnalysisService.kt
package com.mobaassistant.service

import com.mobaassistant.model.Hero
import com.mobaassistant.repository.HeroRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import org.springframework.stereotype.Service
import java.time.Duration
import java.time.Instant

@Service
class MetaAnalysisService(
    private val heroRepository: HeroRepository
) {
    private val metaUpdateInterval = Duration.ofHours(24)
    private var lastUpdate: Instant = Instant.EPOCH
    private var cachedMetaAnalysis: MetaAnalysis? = null

    suspend fun analyzeCurrentMeta(): MetaAnalysis {
        if (shouldUpdateMeta()) {
            cachedMetaAnalysis = performMetaAnalysis()
            lastUpdate = Instant.now()
        }
        return cachedMetaAnalysis ?: performMetaAnalysis()
    }

    fun observeMetaChanges(): Flow<MetaAnalysis> = flow {
        while (true) {
            emit(analyzeCurrentMeta())
            kotlinx.coroutines.delay(metaUpdateInterval.toMillis())
        }
    }

    private fun shouldUpdateMeta(): Boolean {
        return Duration.between(lastUpdate, Instant.now()) > metaUpdateInterval
    }

    private suspend fun performMetaAnalysis(): MetaAnalysis {
        val heroes = heroRepository.findAll()
        return MetaAnalysis(
            topPicks = findTopPicks(heroes),
            banPriorities = findBanPriorities(heroes),
            counterPicks = findCounterPicks(heroes),
            metaStrategies = determineMetaStrategies()
        )
    }

    private fun findTopPicks(heroes: List<Hero>): List<String> {
        return heroes
            .filter { hero -> 
                val stats = getHeroStats(hero)
                stats.pickRate > 0.5 && stats.winRate > 0.5
            }
            .map { it.name }
    }

    private fun findBanPriorities(heroes: List<Hero>): List<String> {
        return heroes
            .filter { hero ->
                val stats = getHeroStats(hero)
                stats.banRate > 0.3
            }
            .map { it.name }
    }

    private fun findCounterPicks(heroes: List<Hero>): Map<String, List<String>> {
        return heroes.associate { hero ->
            hero.name to hero.counters
        }
    }

    private fun determineMetaStrategies(): List<String> {
        return listOf(
            "Early game aggression",
            "Objective-focused gameplay",
            "Split push pressure",
            "Team fight composition"
        )
    }

    private fun getHeroStats(hero: Hero): HeroMetaStats {
        // In a real implementation, this would fetch from a database or API
        return HeroMetaStats(
            pickRate = 0.6,
            winRate = 0.52,
            banRate = 0.3
        )
    }
}

data class MetaAnalysis(
    val topPicks: List<String>,
    val banPriorities: List<String>,
    val counterPicks: Map<String, List<String>>,
    val metaStrategies: List<String>
)

data class HeroMetaStats(
    val pickRate: Double,
    val winRate: Double,
    val banRate: Double
)

File: app/services/meta-analysis.service.ts
import { Observable } from '@nativescript/core';
import { Hero } from '../models/hero.model';
import { MetaAnalysis } from '../models/meta.model';

export class MetaAnalysisService extends Observable {
  private heroStats: Map<string, any> = new Map();
  private readonly META_UPDATE_INTERVAL = 24 * 60 * 60 * 1000; // 24 hours

  async analyzeCurrentMeta(): Promise<MetaAnalysis> {
    await this.updateMetaData();

    return {
      topPicks: this.getTopPicks(),
      banPriorities: this.getBanPriorities(),
      counterPicks: this.getCounterPicks(),
      metaStrategies: this.getMetaStrategies()
    };
  }

  private async updateMetaData(): Promise<void> {
    // In production, fetch from API
    // For now, using static data
    this.heroStats = new Map([
      ['Beatrix', { pickRate: 0.8, winRate: 0.55, banRate: 0.4 }],
      ['Ling', { pickRate: 0.7, winRate: 0.52, banRate: 0.6 }]
    ]);
  }

  private getTopPicks(): string[] {
    const picks: string[] = [];
    this.heroStats.forEach((stats, hero) => {
      if (stats.pickRate > 0.5 && stats.winRate > 0.5) {
        picks.push(hero);
      }
    });
    return picks;
  }

  private getBanPriorities(): string[] {
    const bans: string[] = [];
    this.heroStats.forEach((stats, hero) => {
      if (stats.banRate > 0.3) {
        bans.push(hero);
      }
    });
    return bans;
  }

  private getCounterPicks(): Map<string, string[]> {
    const counters = new Map<string, string[]>();
    // Implement counter pick logic based on hero matchup data
    return counters;
  }

  private getMetaStrategies(): string[] {
    return [
      'Early game aggression',
      'Objective-focused gameplay',
      'Split push pressure',
      'Team fight composition'
    ];
  }
}